File: sdks/python/scripts/cypher_parity.py
"""Run unchanged shared Cypher fixtures through public SDK methods, without retries."""

from __future__ import annotations

import json
import os
import re
from pathlib import Path

from helixdb import AsyncClient, Client, Disk, EmbeddedCacheConfig, HelixError, InMemory


def cases() -> list[dict]:
    path = Path(
        os.environ.get(
            "HELIX_CYPHER_PARITY_FIXTURES",
            Path(__file__).resolve().parents[2] / "tests" / "cypher" / "runtime.json",
        )
    )
    corpus = json.loads(path.read_text(encoding="utf-8"))
    if corpus.get("schema_version") != 1 or len(corpus["cases"]) != 10:
        raise ValueError("unsupported or incomplete Cypher parity corpus")
    seen: set[str] = set()
    for case in corpus["cases"]:
        name = case["name"]
        if not isinstance(name, str) or not re.fullmatch(r"[a-z0-9-]+", name) or name in seen:
            raise ValueError("Cypher parity names must be unique safe basenames")
        seen.add(name)
    return corpus["cases"]


def write_case(client: Client, case: dict, root: Path) -> None:
    try:
        output = {
            "result": client.cypher(
                case["query"],
                case.get("parameters", {}),
                query_name=case["name"],
            )
        }
    except HelixError as error:
        output = {"error": str(error), "code": error.code, "details": error.server_details}
    (root / f"{case['name']}.json").write_text(
        json.dumps(output, ensure_ascii=False, allow_nan=False),
        encoding="utf-8",
    )


async def write_case_async(client: AsyncClient, case: dict, root: Path) -> None:
    try:
        output = {
            "result": await client.cypher(
                case["query"],
                case.get("parameters", {}),
                query_name=case["name"],
            )
        }
    except HelixError as error:
        output = {"error": str(error), "code": error.code, "details": error.server_details}
    (root / f"{case['name']}.json").write_text(
        json.dumps(output, ensure_ascii=False, allow_nan=False),
        encoding="utf-8",
    )


def run_embedded(source: Disk | InMemory, cache: EmbeddedCacheConfig, results: Path) -> None:
    fixtures = cases()
    root = results / "cypher"
    root.mkdir(parents=True, exist_ok=True)
    client = Client.embedded(source, cache=cache)
    try:
        for case in fixtures:
            if isinstance(source, Disk) and case.get("after_disk_reopen", False):
                client.close()
                reader = Client.embedded_reader(source, cache=cache)
                try:
                    write_case(reader, case, root)
                finally:
                    reader.close()
                client = Client.embedded(source, cache=cache)
            else:
                write_case(client, case, root)
    finally:
        client.close()


async def run_embedded_async(
    source: Disk | InMemory, cache: EmbeddedCacheConfig, results: Path
) -> None:
    fixtures = cases()
    root = results / "cypher"
    root.mkdir(parents=True, exist_ok=True)
    client = await AsyncClient.embedded(source, cache=cache)
    try:
        for case in fixtures:
            if isinstance(source, Disk) and case.get("after_disk_reopen", False):
                await client.close()
                reader = await AsyncClient.embedded_reader(source, cache=cache)
                try:
                    await write_case_async(reader, case, root)
                finally:
                    await reader.close()
                client = await AsyncClient.embedded(source, cache=cache)
            else:
                await write_case_async(client, case, root)
    finally:
        await client.close()

File: sdks/rust/examples/support/cypher_parity.rs
//! Real Cypher calls through the public Rust SDK. The common parity checker
//! validates these raw results against the independently authored fixture set.
use helix_db::{CacheConfig, Client, DbConfig, HelixDbSource};
use std::{collections::BTreeMap, fs, path::Path};

#[derive(serde::Deserialize)]
struct Corpus {
    schema_version: u32,
    cases: Vec<Case>,
}

#[derive(serde::Deserialize)]
struct Case {
    name: String,
    query: String,
    #[serde(default)]
    parameters: BTreeMap<String, serde_json::Value>,
    #[serde(default)]
    after_disk_reopen: bool,
}

fn load() -> Result<Vec<Case>, Box<dyn std::error::Error>> {
    let fixture = std::env::var("HELIX_CYPHER_PARITY_FIXTURES").unwrap_or_else(|_| {
        concat!(env!("CARGO_MANIFEST_DIR"), "/../tests/cypher/runtime.json").into()
    });
    let corpus: Corpus = serde_json::from_slice(&fs::read(fixture)?)?;
    if corpus.schema_version != 1 || corpus.cases.len() != 10 {
        return Err("unsupported or incomplete Cypher parity corpus".into());
    }
    let mut names = std::collections::BTreeSet::new();
    for case in &corpus.cases {
        if case.name.is_empty()
            || !case
                .name
                .bytes()
                .all(|byte| byte.is_ascii_alphanumeric() || byte == b'-')
            || !names.insert(&case.name)
        {
            return Err("Cypher parity names must be unique safe basenames".into());
        }
    }
    Ok(corpus.cases)
}

pub async fn run(source: HelixDbSource, results: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let cases = load()?;
    let results = results.join("cypher");
    fs::create_dir_all(&results)?;
    let config = || DbConfig::new().with_cache(CacheConfig::default());
    let mut client = Client::open_with_config(source.clone(), config()).await?;
    let execution = async {
        for case in &cases {
            let response = if case.after_disk_reopen && matches!(source, HelixDbSource::Disk { .. })
            {
                client.close().await?;
                let reader = Client::open_reader_with_config(source.clone(), config()).await?;
                let response = reader
                    .cypher(&case.query, case.parameters.clone(), Some(&case.name))
                    .await;
                reader.close().await?;
                client = Client::open_with_config(source.clone(), config()).await?;
                response
            } else {
                client
                    .cypher(&case.query, case.parameters.clone(), Some(&case.name))
                    .await
            };
            // Record actual failures without substituting the expected detail.
            // Cypher writes are never retried by this fixture driver.
            let response = match response {
                Ok(response) => serde_json::json!({"result": {
                    "columns": response.columns,
                    "rows": response.rows,
                }}),
                Err(error) => serde_json::json!({"error": error.to_string(), "code": error.error_code(), "details": error.remote_details()}),
            };
            fs::write(
                results.join(format!("{}.json", case.name)),
                serde_json::to_vec(&response)?,
            )?;
        }
        Ok::<_, Box<dyn std::error::Error>>(())
    }
    .await;
    let closed = client.close().await;
    execution?;
    closed?;
    Ok(())
}

/// Execute either side of the parent harness's local HTTP server restart.
pub async fn run_http(
    url: &str,
    results: &Path,
    phase: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let cases = load()?;
    let after = match phase {
        "before" => false,
        "after" => true,
        _ => return Err("invalid Cypher HTTP phase".into()),
    };
    let results = results.join("cypher");
    fs::create_dir_all(&results)?;
    let client = Client::server(Some(url))?;
    let execution = async {
        let mut reopened = false;
        for case in cases {
            reopened |= case.after_disk_reopen;
            if reopened != after {
                continue;
            }
            let response = match client
                .cypher(&case.query, case.parameters, Some(&case.name))
                .await
            {
                Ok(response) => serde_json::json!({"result": {
                    "columns": response.columns,
                    "rows": response.rows,
                }}),
                Err(error) => serde_json::json!({"error": error.to_string(), "code": error.error_code(), "details": error.remote_details()}),
            };
            fs::write(
                results.join(format!("{}.json", case.name)),
                serde_json::to_vec(&response)?,
            )?;
        }
        Ok::<_, Box<dyn std::error::Error>>(())
    }
    .await;
    let closed = client.close().await;
    execution?;
    closed?;
    Ok(())
}
